Amazon Provider Package user agent

ferruzzi opened the pull request:
Adding a couple of fields to the boto3 user agent will allow the AWS team to better understand which services and operators to focus improvements on in the future.  This is similar to the user agent fields added by [Databricks](https://github.com/apache/airflow/blob/main/airflow/providers/databricks/hooks/databricks_base.py#L141), [Google](https://github.com/apache/airflow/blob/main/airflow/providers/google/common/hooks/base_google.py#L322), [Yandex](https://github.com/apache/airflow/blob/main/airflow/providers/yandex/hooks/yandex.py#L93), and others.

ferruzzi committed:
Add Caller to User Agent

ferruzzi committed:
Add DAG Key to User Agent

ferruzzi committed:
Add Airflow and AmPP version numbers to User Agent

uranusjr commented on airflow/providers/amazon/aws/hooks/base_aws.py:
```diff
@@ -405,9 +411,68 @@ def __init__(
+    def _get_provider_version(cls) -> str:
+        """Checks the Providers Manager for the package version."""
+        manager = ProvidersManager()
+        provider_name = manager.hooks[cls.conn_type].package_name  # type: ignore[union-attr]
```
What is this ignoring?

ferruzzi replied:
`ProvidersManager().hooks` is an `Optional[HookInfo]`, so MyPy will complain that
 ```
error: Item "None" of "Optional[HookInfo]" has no attribute "package_name"  [union-attr]
```
 
 
 I could do something like this instead.  It feels more confusing, but it does catch the (unlikely??) exception case where ProvidersManager breaks?


```
    @classmethod
    def _get_provider_version(cls) -> str:
        """Checks the Providers Manager for the package version."""
        manager = ProvidersManager()
        hook = manager.hooks[cls.conn_type]
        if not hook:
            return "Unknown"
        provider = manager.providers[hook.package_name]
        return provider.version
```

ferruzzi replied:
Or, following the precedent in `_get_caller`, since this should NEVER cause an issue to bubble up to the user, maybe this is better?

```
    @classmethod
    def _get_provider_version(cls) -> str:
        """Checks the Providers Manager for the package version."""
        try:
            manager = ProvidersManager()
            hook = manager.hooks[cls.conn_type]
            if not hook:
                raise ValueError(f"Hook info for {cls.conn_type} not found in the Provider Manager.")
            provider = manager.providers[hook.package_name]
            return provider.version
        except Exception:
            # While it is generally considered poor form to catch "Exception", under
            # no condition should an error here ever cause an issue for the user.
            return "Unknown"
```

ferruzzi replied:
Both of those solutions feel a bit unnecessarily complicated, but I can make the change if you have a preference.

uranusjr replied:
Checking `if not hook` (or `if hook is None`) makes sense and looks simple enough to me.

ferruzzi replied:
Alright.   I'm grabbing dinner, I'll make the change ~tomorrow~ soon.   Thanks for catching that.

ferruzzi replied:
Sorry for the delay, just got back from vacation and it took a little longer to get back into gear.  I ended up wrapping it in a try/except as mentioned and dropped the `if not hook`.  If `hook` is falsy, then it'll error out on the next line at `hook.package_name` anyway and get caught by the `except`.

Addressed in https://github.com/apache/airflow/pull/27823/commits/a5fc3bc4a39855b9f3c7fc5ac26505709d191a98

Taragolis commented on airflow/providers/amazon/aws/hooks/base_aws.py:
```diff
@@ -405,9 +411,68 @@ def __init__(
+        # That salted value is used to generate a SHA-1 hash which, by definition,
+        # can not (reasonably) be reversed.  No personal data can be inferred or
+        # extracted from the resulting UUID.
+        return str(uuid.uuid5(uuid.NAMESPACE_OID, dag_id))
```
Just an idea, might if it run DAG_ID context not set then return NIL UUID `00000000-0000-0000-0000-000000000000`?

ferruzzi replied:
Yeah, that's reasonable.  To be honest, it doesn't really matter.  A hash of two known/fixed values would make a known fixed value, but I suppose all zeroes would be even more obvious.  If I change the provider version to return "Unknown" per the discussion above with [uranusjr](https://github.com/uranusjr), then maybe having this default to "Unknown" or all zeroes is a better option as well.

Taragolis commented on airflow/providers/amazon/aws/hooks/base_aws.py:
```diff
@@ -42,11 +46,13 @@
 from dateutil.tz import tzlocal
 from slugify import slugify
 
+from airflow import __version__ as airflow_version
```
@ferruzzi find also new circular imports, same as previously fixed in #26784

It is only happen if the end user stored secrets configuration in secrets backend.
And unfortunetly it can't be unit tested (in general way) because during unit tests most of the imports already imported

```python
❯ export AIRFLOW__SECRETS__BACKEND=airflow.providers.amazon.aws.secrets.secrets_manager.SecretsManagerBackend
❯ export AIRFLOW__SECRETS__BACKEND_KWARGS='{"config_prefix": "/airflow/config", "region_name": "eu-west-1"}'
❯ export AIRFLOW__DATABASE__SQL_ALCHEMY_CONN_SECRET=boom
❯ airflow version
Traceback (most recent call last):
  File "/Users/taragolis/Projects/common/airflow/airflow/providers/amazon/aws/secrets/secrets_manager.py", line 449, in _get_secret
    response = self.client.get_secret_value(
  File "/Users/taragolis/.pyenv/versions/3.9.10/lib/python3.9/functools.py", line 993, in __get__
    val = self.func(instance)
  File "/Users/taragolis/Projects/common/airflow/airflow/providers/amazon/aws/secrets/secrets_manager.py", line 170, in client
    from airflow.providers.amazon.aws.hooks.base_aws import SessionFactory
  File "/Users/taragolis/Projects/common/airflow/airflow/providers/amazon/aws/hooks/base_aws.py", line 49, in <module>
    from airflow import __version__ as airflow_version
ImportError: cannot import name '__version__' from partially initialized module 'airflow' (most likely due to a circular import) (/Users/taragolis/Projects/common/airflow/airflow/__init__.py)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/taragolis/Projects/common/airflow/airflow/configuration.py", line 123, in _get_config_value_from_secret_backend
    return secrets_client.get_config(config_key)
  File "/Users/taragolis/Projects/common/airflow/airflow/providers/amazon/aws/secrets/secrets_manager.py", line 434, in get_config
    return self._get_secret(self.config_prefix, key)
  File "/Users/taragolis/Projects/common/airflow/airflow/providers/amazon/aws/secrets/secrets_manager.py", line 453, in _get_secret
    except self.client.exceptions.ResourceNotFoundException:
  File "/Users/taragolis/.pyenv/versions/3.9.10/lib/python3.9/functools.py", line 993, in __get__
    val = self.func(instance)
  File "/Users/taragolis/Projects/common/airflow/airflow/providers/amazon/aws/secrets/secrets_manager.py", line 170, in client
    from airflow.providers.amazon.aws.hooks.base_aws import SessionFactory
  File "/Users/taragolis/Projects/common/airflow/airflow/providers/amazon/aws/hooks/base_aws.py", line 49, in <module>
    from airflow import __version__ as airflow_version
ImportError: cannot import name '__version__' from partially initialized module 'airflow' (most likely due to a circular import) (/Users/taragolis/Projects/common/airflow/airflow/__init__.py)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/taragolis/.pyenv/versions/3.9.10/envs/airflow-dev-env-39/bin/airflow", line 33, in <module>
    sys.exit(load_entry_point('apache-airflow', 'console_scripts', 'airflow')())
  File "/Users/taragolis/.pyenv/versions/3.9.10/envs/airflow-dev-env-39/bin/airflow", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "/Users/taragolis/.pyenv/versions/3.9.10/lib/python3.9/importlib/metadata.py", line 77, in load
    module = import_module(match.group('module'))
  File "/Users/taragolis/.pyenv/versions/3.9.10/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 972, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/Users/taragolis/Projects/common/airflow/airflow/__init__.py", line 34, in <module>
    from airflow import settings
  File "/Users/taragolis/Projects/common/airflow/airflow/settings.py", line 37, in <module>
    from airflow.configuration import AIRFLOW_HOME, WEBSERVER_CONFIG, conf  # NOQA F401
  File "/Users/taragolis/Projects/common/airflow/airflow/configuration.py", line 1685, in <module>
    conf.validate()
  File "/Users/taragolis/Projects/common/airflow/airflow/configuration.py", line 349, in validate
    self._validate_config_dependencies()
  File "/Users/taragolis/Projects/common/airflow/airflow/configuration.py", line 449, in _validate_config_dependencies
    is_sqlite = "sqlite" in self.get("database", "sql_alchemy_conn")
  File "/Users/taragolis/Projects/common/airflow/airflow/configuration.py", line 578, in get
    option = self._get_environment_variables(deprecated_key, deprecated_section, key, section)
  File "/Users/taragolis/Projects/common/airflow/airflow/configuration.py", line 657, in _get_environment_variables
    option = self._get_env_var_option(section, key)
  File "/Users/taragolis/Projects/common/airflow/airflow/configuration.py", line 502, in _get_env_var_option
    return _get_config_value_from_secret_backend(os.environ[env_var_secret_path])
  File "/Users/taragolis/Projects/common/airflow/airflow/configuration.py", line 125, in _get_config_value_from_secret_backend
    raise AirflowConfigException(
airflow.exceptions.AirflowConfigException: Cannot retrieve config from alternative secrets backend. Make sure it is configured properly and that the Backend is accessible.
cannot import name '__version__' from partially initialized module 'airflow' (most likely due to a circular import) (/Users/taragolis/Projects/common/airflow/airflow/__init__.py)
```

ferruzzi replied:
I think I addressed this in https://github.com/apache/airflow/pull/27823/commits/a5fc3bc4a39855b9f3c7fc5ac26505709d191a98 by moving it to a local import in the helper method.

Taragolis replied:
Yep, this should help. Secrets Backends do not use Hook for communicate with SSM Parameter Store / Secrets Manager: https://github.com/apache/airflow/pull/27823#discussion_r1029801751

ferruzzi commented on airflow/providers/amazon/aws/hooks/base_aws.py:
```diff
@@ -25,9 +25,13 @@
 from __future__ import annotations
 
```
Moving to code comments so these can be threaded discussions:

@Taragolis :
> Someone of end users might be unhappy the fact that we actually started collect additional telemetry like.
dag_id and ClassName even if it only hashes. So might be better collect by default only Airflow and Amazon Provider version and if required additional metadata then call optional callable which could controlled by config e.g. [aws] extra_botocore_user_agent_callable with full qualified path to callable e.g. some_vendor.safe.get_user_agent

Yeah, it is possible that this may raise an eyebrow and that is something I put a lot of thought into.  In the end I decided that Databricks is already adding the calling method to their user agent field [see link in the PR description] so I felt it was already approved by the community.  And as you mentioned, the Dag ID is a hashed value that never contained any kind of secret/personal/private information even if it could somehow be "decoded".  I am definitely open to hear what others think, but I think we're safe here (obviously, or I wouldn't have submitted it)

In the long term I was thinking it would be neat to try to consolidate these various user agent collection/building methods that various different providers are doing and make some kind of contribution to core that vends these values.  That would make the "let the user opt in to what they are willing to share" concept much easier.  They could easily see what they are sharing universally across providers in one place, and providers who wish to add these values to their agent don't all have to rewrite the book.   I basically copy/pasted the "get package version" method from Yandex [link in PR description] and if another provider wants to include it in their user agent, that's yet another bowl of copypasta.  But a central provider/builder is outside the scope of this PR.  At least three other providers are doing this their own way already, I'd like to get AWS added to that before we start trying to sort out how to standardize everything across providers.

Taragolis replied:
About calling method/function I mean something like that (just simplified version)

```python
from airflow.configuration import conf


def user_agent() -> dict:
    # Construct some generic part, like airflow version and provider version
    ua = {
        "airflow": "3.0.0",
        "amazon-provider": "15.0.0"
    }
    
    fn = conf.getimport("aws", "extra_botocore_user_agent_callable", fallback=None)
    if fn and callable(fn):
        # Vendor user-agent
        ua.update(fn())
    
    return ua
```

Something similar that snowflake provider has

https://github.com/apache/airflow/blob/035315f5cf331a0fc5b1800f708e2f83fcfb2d7a/airflow/providers/snowflake/hooks/snowflake.py#L233-L234

shubham22 replied:
>     1. Someone of end users might be unhappy the fact that we actually started collect additional telemetry like.
>        dag_id and ClassName even if it only hashes. So might be better collect by default only Airflow and Amazon Provider version and if required additional metadata then call optional callable which could controlled by config e.g. `[aws] extra_botocore_user_agent_callable` with full qualified path to callable e.g. `some_vendor.safe.get_user_agent`

While I understand where you're coming from, I disagree with this on multiple aspects - 
1. Hashed Id doesn't really give us any information about what a user could be naming. It is completely private, and doesn't incur any cost. It rather allows us to make better decisions on our end to serve the users better. If it wasn't hashed, I would agree with your suggestion.

2. No more additional configs, please. We already give a lot of configs to the users, and we should try to minimize the overuse.

Taragolis replied:
1. I can't say that hashes with known algorithm and salt is "completely private". It can't be easily revert however can be easily identify if has a list of known dags, e.g. "5ee762d8-7bac-5131-9887-03dcab3b7111" stands for "example_s3".
We could talk about this kind of stuff for ages and the result would be that more people would like make/buy "tin foil hat".

2. This config intend to use by vendors of Airflow-as-a-service not end users it also could be not a general Airflow Config it might be just an environment variable "AIRFLOW_BOTOCORE_ADDITIONAL_UA". So this callable could return any vendor specific info, for example:
   - Version of MWAA
   - Version of local runner
   - Version of Google Composer
   - Version of Astronomer
   - Any DAG/Callable related stuff

shubham22 replied:
1. I agree that if you already know a list of known dags, you can figure it out. If there is a better suggestion that makes salt unknown but keeps it same across the DAG runs, I think that would be preferred.

2. I like the idea that we are not adding this config on the user. But, does it solve the problem we started with i.e. providing more privacy to the users? Would we end up adding this config for every provider out there? I still think it is not needed.

Taragolis replied:
I just say about my ideas and suggestion If we talk about me I'm a person who usual check "Send Anonymous Statistic to <CompanyName> for improve <Software Name>" and allow all cookies on every website 😆 .

In general it might be a problem to decide what is actually PII data and what allowed to collect and usual it regulate by different way in different Countries / Regions. 

So idea if some additional data for user agent required than vendor decide what should be enabled depend on endusers regions. It just my assumption because I'm not a lawyer

shubham22 replied:
Hahaha really appreciate the help here. Let me consult with our legal counsel to ensure that we are not crossing any legal boundaries here with respect to particular countries/regions :)

And I haven't met a person who doesn't click "accept all cookies" yet. No idea what purpose that annoying as hell pop-up ever achieved.

ferruzzi replied:
@shubham22 @Taragolis  - Did you guys work this out?  Can I resolve this thread or is there still some discussion to work through here?

shubham22 replied:
I confirmed with the managed service legal counsel that we aren't crossing any legal boundaries here. I think we can close it.

ferruzzi replied:
cool, thanks @shubham22 

@Taragolis  - Are we good to proceed?

Taragolis replied:
No concern from my side

ferruzzi commented on airflow/providers/amazon/aws/hooks/base_aws.py:
```diff
@@ -25,9 +25,13 @@
 from __future__ import annotations
 
```
> I thought we also might add/merge config in AwsCredentialHelper rather than in hook

@Taragolis  What would be the advantage of adding it there instead?

Taragolis replied:
Right now AwsCredentialHelper resolve connection configurations between Hook args and Connection parameters.
So I assume that a good place to keep it there

Taragolis replied:
Ohhh... and I just figure out that amazon secrets backends do not use `AwsBaseHook` and directly use helper and session factory

ferruzzi replied:
Sorry for the wait, I wanted to think this through and make sure I understood the implications.  If I am understanding correctly: doing it the way it is currently written means that we do not get any data from when Airflow itself connects to SecretsManager.  That was somewhat accidental, but I consider that an absolute win.  I/We don't want to be "snooping" on traffic where Airflow is retrieving secrets.  The main goal here is to understand which services and endpoints are being used in DAGs so we can prioritize adding new hooks and operators accordingly.

ferruzzi replied:
@Taragolis What do you think?   Is this alright or do you still want that change?

Taragolis replied:
@ferruzzi Sorry miss notification.

This more about preventing circular imports. Right now if user try to obtain _config_  from Secret Backend ([with `_secret` suffix](https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html#setting-configuration-options)) it actually happen before some airflow modules properly imported and in this case local import might not help.

It should not affect if user get _variables_ and _connections_ from Secrets Backends.

That mean right now it is fine that this imports inside methods of AwsGenericHook/AwsBaseHook

ferruzzi committed:
PR Fixes

- Ensure all helper methods have a full `except Exception` block with reasonable default values
- Moved possible circular import into a helper within a try/except block

uranusjr reviewed (approved):
Pending the discussion on telemetry

potiuk reviewed (approved):
Some static checks need fixing, but otherwise it looks good.

ferruzzi committed:
MyPy fixes

ferruzzi committed:
import order fix

ferruzzi committed:
Fix config default overwriting user values

Taragolis commented on airflow/providers/amazon/aws/hooks/base_aws.py:
```diff
@@ -409,6 +414,92 @@ def __init__(
+            return str(uuid.uuid5(uuid.NAMESPACE_OID, dag_id))
+        except Exception:
+            # Under no condition should an error here ever cause an issue for the user.
+            return "00000000-0000-5000-0000-000000000000"
```
Small nitpick 

1. Seems like there is only KeyError might happen here
2. Some fixes in nil-uuid



```suggestion
        except KeyError:
            # Under no condition should an error here ever cause an issue for the user.
            return "00000000-0000-0000-0000-000000000000"
```

ferruzzi replied:
The 5 is a fixed bit in UUID to show the version and the unit tests check for it to confirm the format so that either has to stay or the unit tests need to be changed.  I'd like to keep it, but if you have a strong opinion here or a reason it should be changed I'll update the unit tests too.

For the exception, yeah you are right but I did it for consistency since the others all go by the policy that "nothing should possibly bubble up".  I can change it to IndexError if you still want, I just figured I would explain my reason.

Taragolis replied:
Nil UUID is special form of UUID, it is not include any specific data, see: https://datatracker.ietf.org/doc/html/rfc4122.html#section-4-1-7

So it is good point to check if any exception happen then all bits are 0

```python
from uuid import UUID

nil = "00000000-0000-0000-0000-000000000000"
assert UUID(int=0) == UUID(nil)
```

Taragolis replied:
And actually I thought we do not need any regex. We could just pass value to UUID object and compare with version

```python
from uuid import UUID

random = "80eb85a0-5aef-45b6-abbb-f16d62d3db42"
uuid_v5 = "bf428e1d-f221-55de-a77f-a61755a4d727"
nil = "00000000-0000-0000-0000-000000000000"

assert UUID(random).version == 4
assert UUID(uuid_v5).version == 5
assert UUID(nil).version is None
```

And nil uuid not possible to get by use uuid5 however in theory (and infinity time) it is possible to get `00000000-0000-5000-0000-000000000000`

ferruzzi replied:
I'm just running the static checks locally and I'll push the no-regex version.  I actually didn't know about the UUID().version check, that's very handy. :+1:

Taragolis replied:
I didn't know that one of the bit is actually a version until you told that `¯\_(ツ)_/¯`  👍

ferruzzi replied:
Alright, I made a small tweak and rerunning the tests locally, but what I ended up with is `assert UUID(dag_run_key).version in {5, None}`.  UUID().version also verifies that it is a valid format so that will catch a poorly formed UUID or anything thast is a valid UUID but not v5 or nil

ferruzzi replied:
and then realized that without mocking the environment variable that generated the UUID, it's always returning the exception case so I've parameterized the test so it's testing both cases.  Tests should be done in a sec.

ferruzzi committed:
Fix UUID checks
